include/nbl/video/ILogicalDevice.h

```diff
@@ -539,19 +543,133 @@ class ILogicalDevice : public core::IReferenceCounted
 
     virtual void waitIdle() = 0;
 
+    virtual void* mapMemory(const IDriverMemoryAllocation::MappedMemoryRange& memory, IDriverMemoryAllocation::E_MAPPING_CPU_ACCESS_FLAG accessHint = IDriverMemoryAllocation::EMCAF_READ_AND_WRITE) = 0;
+
+    virtual void unmapMemory(IDriverMemoryAllocation* memory) = 0;
+
+    inline void updateBufferRangeViaStagingBuffer(IGPUCommandBuffer* cmdbuf, IGPUFence* fence, IGPUQueue* queue, const asset::SBufferRange<IGPUBuffer>& bufferRange, const void* data)
+    {
+        auto* cmdpool = cmdbuf->getPool();
+        assert(cmdpool->getCreationFlags() & IGPUCommandPool::ECF_RESET_COMMAND_BUFFER_BIT);
+        assert(cmdpool->getQueueFamilyIndex() == queue->getFamilyIndex());
+
+        //EventHandle event = null;
+        resetFences(1u, &fence);
+        for (size_t uploadedSize=0ull; uploadedSize<bufferRange.size;)
+        {
+            const void* dataPtr = reinterpret_cast<const uint8_t*>(data) + uploadedSize;
+            uint32_t localOffset = video::StreamingTransientDataBufferMT<>::invalid_address;
+            uint32_t alignment = 64u; // smallest mapping alignment capability
+            uint32_t subSize = static_cast<uint32_t>(core::min<uint32_t>(core::alignDown(m_defaultUploadBuffer.get()->max_size(), alignment), bufferRange.size - uploadedSize));
+            m_defaultUploadBuffer.get()->multi_place(std::chrono::high_resolution_clock::now() + std::chrono::microseconds(500u), 1u, (const void* const*)&dataPtr, &localOffset, &subSize, &alignment);
+
+            // keep trying again
+            if (localOffset == video::StreamingTransientDataBufferMT<>::invalid_address)
+            {
+                // but first sumbit the already buffered up copies
+                cmdbuf->end();
+                IGPUQueue::SSubmitInfo submit;
+                submit.commandBufferCount = 1u;
+                submit.commandBuffers = &cmdbuf;
+                submit.signalSemaphoreCount = 0u;
+                submit.pSignalSemaphores = nullptr;
+                submit.waitSemaphoreCount = 0u;
+                submit.pWaitSemaphores = nullptr;
+                submit.pWaitDstStageMask = nullptr;
```

Comment:
question to resolve... even if execution dependency is maintained (by having the cpu wait till submit is done before carrying on to the next one), do we need a memory barrier?

Reply:
will check tomorrow, can't see what part of code is it about XD

Reply:
basically if you have
```
submit A
wait fence signalled by A
submit B
```
can submit B stomp on submit A (probably not because A needs to complete before the host can proceed to submit B, and B cannot start before its submitted)

However there's a question whether all memory operations from A (scopes are important) will be visible to the host and submit B

Reply:
basically its a question whether there needs to be extra sync despite the fence, such as a pipeline barrier/event if A and B are submitted to the same queue even if host waits on a fence between them